False positive with anonymous class

It appears that constructing an anonymous instance with a nullable parameter doesn't get inspected correctly.

```
/.../AbstractLinkedDeque.java:346: error: [NullAway] passing @Nullable parameter 'first'
where @NonNull is required
    return new AbstractLinkedIterator(first) {
                                      ^
```

```java
public PeekingIterator<E> iterator() {
  return new AbstractLinkedIterator(first) {
    @Override @Nullable E computeNext() {
      return getNext(cursor);
    }
  };
}

abstract class AbstractLinkedIterator implements PeekingIterator<E> {
  @Nullable E previous;
  @Nullable E cursor;

  AbstractLinkedIterator(@Nullable E start) {
    cursor = start;
  }
  ...
}
```

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -1878,9 +1878,8 @@ public class NullAway extends BugChecker
     Scope scope = superType.tsym.members();
     for (Symbol sym : scope.getSymbolsByName(methodSymbol.name)) {
       if (sym != null
-          && !sym.isStatic()
+          && sym.isConstructor()
           && ((sym.flags() & Flags.SYNTHETIC) == 0)
-          && sym.name.contentEquals(methodSymbol.name)
           && hasSameArgTypes((Symbol.MethodSymbol) sym, methodSymbol, types)) {
         return (Symbol.MethodSymbol) sym;
       }

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -32,6 +32,7 @@ import static com.sun.source.tree.Tree.Kind.TYPE_CAST;
 import com.google.auto.service.AutoService;
 import com.google.auto.value.AutoValue;
 import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMultimap;
 import com.google.common.collect.ImmutableSet;
@@ -78,6 +79,8 @@ import com.sun.source.tree.VariableTree;
 import com.sun.source.tree.WhileLoopTree;
 import com.sun.source.util.TreePath;
 import com.sun.source.util.Trees;
+import com.sun.tools.javac.code.Flags;
+import com.sun.tools.javac.code.Scope;
 import com.sun.tools.javac.code.Symbol;
 import com.sun.tools.javac.code.Symbol.VarSymbol;
 import com.sun.tools.javac.code.Type;
@@ -302,11 +305,22 @@ public class NullAway extends BugChecker
     if (!matchWithinClass) {
       return Description.NO_MATCH;
     }
-    final Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(tree);
+    Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(tree);
     if (methodSymbol == null) {
       throw new RuntimeException("not expecting unresolved method here");
     }
     List<? extends ExpressionTree> actualParams = tree.getArguments();
+    if (tree.getClassBody() != null && actualParams.size() > 0) {
+      // passing parameters to constructor of anonymous class
+      // this constructor just invokes the constructor of the superclass, and
+      // in the AST does not have the parameter nullability annotations from the superclass.
+      // so, treat as if the superclass constructor is being invoked directly
+      // see https://github.com/uber/NullAway/issues/102
+      Type supertype = state.getTypes().supertype(methodSymbol.owner.type);
+      Symbol.MethodSymbol superConstructor =
+          findSuperConstructorInType(methodSymbol, supertype, state.getTypes());
+      methodSymbol = superConstructor;
+    }
     return handleInvocation(state, methodSymbol, actualParams);
   }
 
@@ -1850,7 +1864,43 @@ public class NullAway extends BugChecker
     return message;
   }
 
-  public static enum MessageTypes {
+  /**
+   * based heavily on {@link ASTHelpers#findSuperMethodInType(Symbol.MethodSymbol, Type, Types)},
+   * but works for constructors
+   */
+  @Nullable
+  private static Symbol.MethodSymbol findSuperConstructorInType(
+      Symbol.MethodSymbol methodSymbol, Type superType, Types types) {
+    Preconditions.checkArgument(methodSymbol.isConstructor(), "only accepts constructor methods");
+    Scope scope = superType.tsym.members();
+    for (Symbol sym : scope.getSymbolsByName(methodSymbol.name)) {
+      if (sym != null
+          && !sym.isStatic()
+          && ((sym.flags() & Flags.SYNTHETIC) == 0)
+          && sym.name.contentEquals(methodSymbol.name)
+          && hasSameArgTypes((Symbol.MethodSymbol) sym, methodSymbol, types)) {
+        return (Symbol.MethodSymbol) sym;
+      }
+    }
+    return null;
+  }
+
+  private static boolean hasSameArgTypes(
+      Symbol.MethodSymbol method1, Symbol.MethodSymbol method2, Types types) {
+    com.sun.tools.javac.util.List<VarSymbol> method1params = method1.getParameters();
+    com.sun.tools.javac.util.List<VarSymbol> method2params = method2.getParameters();
+    if (method1params.size() != method2params.size()) {
+      return false;
+    }
+    for (int i = 0; i < method1params.size(); i++) {
+      if (!types.isSameType(method1params.get(i).type, method2params.get(i).type)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public enum MessageTypes {
     DEREFERENCE_NULLABLE,
     RETURN_NULLABLE,
     PASS_NULLABLE,

--- a/nullaway/src/main/java/com/uber/nullaway/NullAway.java
+++ b/nullaway/src/main/java/com/uber/nullaway/NullAway.java
@@ -319,6 +319,9 @@ public class NullAway extends BugChecker
       Type supertype = state.getTypes().supertype(methodSymbol.owner.type);
       Symbol.MethodSymbol superConstructor =
           findSuperConstructorInType(methodSymbol, supertype, state.getTypes());
+      if (superConstructor == null) {
+        throw new RuntimeException("must find constructor in supertype");
+      }
       methodSymbol = superConstructor;
     }
     return handleInvocation(state, methodSymbol, actualParams);

--- a/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNegativeCases.java
+++ b/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNegativeCases.java
@@ -738,4 +738,13 @@ public class NullAwayNegativeCases {
       klazz.hashCode();
     }
   }
+
+  static class TestAnon {
+
+    TestAnon(@Nullable Object p) {}
+  }
+
+  static TestAnon testAnon(@Nullable Object q) {
+    return new TestAnon(q) {};
+  }
 }

--- a/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayPositiveCases.java
+++ b/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayPositiveCases.java
@@ -435,4 +435,14 @@ public class NullAwayPositiveCases {
       return null;
     }
   }
+
+  static class TestAnon {
+
+    TestAnon(Object p) {}
+  }
+
+  static TestAnon testAnon(@Nullable Object q) {
+    // BUG: Diagnostic contains: passing @Nullable parameter 'q' where
+    return new TestAnon(q) {};
+  }
 }

